"file too short" error in FFT callback

```
  File "/work/cupy/cupy/testing/helper.py", line 47, in _call_func
    result = impl(self, *args, **kw)
  File "/work/cupy/tests/cupy_tests/fft_tests/test_callback.py", line 134, in test_fft_load
    return self._test_load_helper(xp, dtype, 'fft')
  File "/work/cupy/tests/cupy_tests/fft_tests/test_callback.py", line 126, in _test_load_helper
    with xp.fft.config.set_cufft_callbacks(cb_load=cb_load):
  File "cupy/fft/_callback.pyx", line 545, in cupy.fft._callback.set_cufft_callbacks.__init__
    mgr = _CallbackManager(
  File "cupy/fft/_callback.pyx", line 368, in cupy.fft._callback._CallbackManager.__init__
    self.mod = module = importlib.util.module_from_spec(spec)
  File "<frozen importlib._bootstrap>", line 571, in module_from_spec
  File "<frozen importlib._bootstrap_external>", line 922, in create_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
/home/jenkins-external-slave/.chainer-test/.cupy/cupy_callback_3178b0c73746b8908da0e73b3d1b5361.cpython-36m-x86_64-linux-gnu.so: file too short
```


https://jenkins.preferred.jp/job/chainer/job/daily_master_cupy-shuffled/420/ID=8,label=mn1-p100/testReport/junit/tests.cupy_tests.fft_tests.test_callback/Test1dCallbacks/test_fft_load__param_0__n_None__norm_None__shape__10__7___/

<details>

`tests.cupy_tests.fft_tests.test_callback.Test1dCallbacks.test_fft_load[_param_0_{n=None, norm=None, shape=(10, 7)}]`          (from pytest)        

### Error Message

```
AssertionError: Only cupy raises error    File "/work/cupy/cupy/testing/helper.py", line 47, in _call_func     result = impl(self, *args, **kw)   File "/work/cupy/tests/cupy_tests/fft_tests/test_callback.py", line 134, in test_fft_load     return self._test_load_helper(xp, dtype, 'fft')   File "/work/cupy/tests/cupy_tests/fft_tests/test_callback.py", line 126, in _test_load_helper     with xp.fft.config.set_cufft_callbacks(cb_load=cb_load):   File "cupy/fft/_callback.pyx", line 545, in cupy.fft._callback.set_cufft_callbacks.__init__     mgr = _CallbackManager(   File "cupy/fft/_callback.pyx", line 368, in cupy.fft._callback._CallbackManager.__init__     self.mod = module = importlib.util.module_from_spec(spec)   File "<frozen importlib._bootstrap>", line 571, in module_from_spec   File "<frozen importlib._bootstrap_external>", line 922, in create_module   File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed /home/jenkins-external-slave/.chainer-test/.cupy/cupy_callback_3178b0c73746b8908da0e73b3d1b5361.cpython-36m-x86_64-linux-gnu.so: file too short
```

### Stacktrace

```
self = <<cupy_tests.fft_tests.test_callback.Test1dCallbacks object at 0x7f95d9535940>  parameter: {'n': None, 'norm': None, 'shape': (10, 7)}>
xp = <module 'cupy' from '/work/cupy/cupy/__init__.py'>
dtype = <class 'numpy.complex64'>

    @testing.for_complex_dtypes()
    @testing.numpy_cupy_allclose(rtol=1e-4, atol=1e-7, contiguous_check=False)
    def test_fft_load(self, xp, dtype):
>       return self._test_load_helper(xp, dtype, 'fft')

dtype      = <class 'numpy.complex64'>
self       = <<cupy_tests.fft_tests.test_callback.Test1dCallbacks object at 0x7f95d9535940>  parameter: {'n': None, 'norm': None, 'shape': (10, 7)}>
xp         = <module 'cupy' from '/work/cupy/cupy/__init__.py'>

tests/cupy_tests/fft_tests/test_callback.py:134: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/cupy_tests/fft_tests/test_callback.py:126: in _test_load_helper
    with xp.fft.config.set_cufft_callbacks(cb_load=cb_load):
cupy/fft/_callback.pyx:545: in cupy.fft._callback.set_cufft_callbacks.__init__
    mgr = _CallbackManager(
cupy/fft/_callback.pyx:368: in cupy.fft._callback._CallbackManager.__init__
    self.mod = module = importlib.util.module_from_spec(spec)
<frozen importlib._bootstrap>:571: in module_from_spec
    ???
<frozen importlib._bootstrap_external>:922: in create_module
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

f = <built-in function create_dynamic>
args = (ModuleSpec(name='cupy_callback_3178b0c73746b8908da0e73b3d1b5361', loader=<_frozen_importlib_external.ExtensionFileLoa...s-external-slave/.chainer-test/.cupy/cupy_callback_3178b0c73746b8908da0e73b3d1b5361.cpython-36m-x86_64-linux-gnu.so'),)
kwds = {}

>   ???
E   AssertionError: Only cupy raises error
E   
E     File "/work/cupy/cupy/testing/helper.py", line 47, in _call_func
E       result = impl(self, *args, **kw)
E     File "/work/cupy/tests/cupy_tests/fft_tests/test_callback.py", line 134, in test_fft_load
E       return self._test_load_helper(xp, dtype, 'fft')
E     File "/work/cupy/tests/cupy_tests/fft_tests/test_callback.py", line 126, in _test_load_helper
E       with xp.fft.config.set_cufft_callbacks(cb_load=cb_load):
E     File "cupy/fft/_callback.pyx", line 545, in cupy.fft._callback.set_cufft_callbacks.__init__
E       mgr = _CallbackManager(
E     File "cupy/fft/_callback.pyx", line 368, in cupy.fft._callback._CallbackManager.__init__
E       self.mod = module = importlib.util.module_from_spec(spec)
E     File "<frozen importlib._bootstrap>", line 571, in module_from_spec
E     File "<frozen importlib._bootstrap_external>", line 922, in create_module
E     File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
E   /home/jenkins-external-slave/.chainer-test/.cupy/cupy_callback_3178b0c73746b8908da0e73b3d1b5361.cpython-36m-x86_64-linux-gnu.so: file too short

args       = (ModuleSpec(name='cupy_callback_3178b0c73746b8908da0e73b3d1b5361', loader=<_frozen_importlib_external.ExtensionFileLoa...s-external-slave/.chainer-test/.cupy/cupy_callback_3178b0c73746b8908da0e73b3d1b5361.cpython-36m-x86_64-linux-gnu.so'),)
f          = <built-in function create_dynamic>
kwds       = {}

<frozen importlib._bootstrap>:219: AssertionError
```

### Standard Output

```
dtype is <class 'numpy.complex64'>
```

### Standard Error

```
pytest-cov: Failed to setup subprocess coverage. Environ: {'COV_CORE_SOURCE': ':', 'COV_CORE_CONFIG': '/work/cupy/.coveragerc', 'COV_CORE_DATAFILE': '/work/cupy/.coverage'} Exception: ModuleNotFoundError("No module named 'Cython'",)
```

--- a/cupy/fft/_callback.pyx
+++ b/cupy/fft/_callback.pyx
@@ -176,7 +176,7 @@ cdef inline void _cythonize(str tempdir, str mod_name) except*:
                         '-E', 'CUPY_CUFFT_STATIC=True',
                         os.path.join(tempdir, mod_name + '.pyx'),
                         '-o', os.path.join(tempdir, mod_name + '.cpp')],
-                       env=os.environ)
+                       env=os.environ, cwd=tempdir)
     p.check_returncode()
 
 
@@ -190,7 +190,7 @@ cdef inline void _mod_compile(str tempdir, str mod_name, str obj_host) except*:
                        '-fPIC', '-O2', '-std=c++11',
                        '-c', os.path.join(tempdir, mod_name + '.cpp'),
                        '-o', obj_host],
-                       env=os.environ)
+                       env=os.environ, cwd=tempdir)
     p.check_returncode()
 
 
@@ -206,7 +206,8 @@ cdef inline str _prune(str temp_dir, str cache_dir, str _cufft_ver, str arch):
                                         'lib' + cufft_lib_pruned + '.a')
         if not os.path.isfile(cufft_lib_cached):
             p = subprocess.run([_nvprune, '-arch=sm_' + arch,
-                                cufft_lib_full, '-o', cufft_lib_temp])
+                                cufft_lib_full, '-o', cufft_lib_temp],
+                               env=os.environ, cwd=temp_dir)
             p.check_returncode()
             # atomic move with the destination guaranteed to be overwritten;
             # using os.replace() is also ok here
@@ -244,7 +245,7 @@ cdef inline void _nvcc_compile(
     if cb_store:
         cmd.append('-DHAS_STORE_CALLBACK')
     p = subprocess.run(cmd + ['-o', obj_dev],
-                       env=os.environ,
+                       env=os.environ, cwd=tempdir,
                        stdout=subprocess.PIPE,
                        stderr=subprocess.PIPE)
     try:
@@ -277,7 +278,7 @@ cdef inline void _nvcc_link(
         cmd.append('-lcufft_static')
     cmd += ['-lculibos', '-lpthread', '-o', mod_temp]
 
-    p = subprocess.run(cmd, env=os.environ)
+    p = subprocess.run(cmd, env=os.environ, cwd=tempdir)
     p.check_returncode()
     # atomic move with the destination guaranteed to be overwritten;
     # using os.replace() is also ok here

--- a/cupy/fft/_callback.pyx
+++ b/cupy/fft/_callback.pyx
@@ -108,17 +108,21 @@ def lock_directory(cache_dir, timeout=10, poll_interval=0.05):
         os.close(fd)
 
     t_start = time.time()
-    while True:
-        fd = acquire(lock_file)
+    fd = None
+    try:
+        while True:
+            fd = acquire(lock_file)
+            if fd is not None:
+                yield
+                break
+            elif time.time() - t_start > timeout:
+                raise RuntimeError('timeout is reached but a file lock still '
+                                   'cannot be acquired')
+            else:
+                time.sleep(poll_interval)
+    finally:  # exceptions can be raised from yield
         if fd is not None:
-            yield
-            break
-        elif time.time() - t_start > timeout:
-            raise RuntimeError('timeout is reached but a file lock still '
-                               'cannot be acquired')
-        else:
-            time.sleep(poll_interval)
-    release(fd)
+            release(fd)
 
 
 cdef inline void _set_vars() except*:

--- a/cupy/fft/_callback.pyx
+++ b/cupy/fft/_callback.pyx
@@ -5,8 +5,6 @@ from cupy_backends.cuda.api.runtime cimport _is_hip_environment
 from cupy.core.core cimport ndarray
 from cupy.cuda.device cimport get_compute_capability
 
-import contextlib
-import fcntl
 import hashlib
 import importlib
 import os
@@ -17,7 +15,6 @@ import sys
 import sysconfig
 import tempfile
 import threading
-import time
 import warnings
 
 from cupy import __version__ as _cupy_ver
@@ -71,58 +68,6 @@ cdef class _ThreadLocal:
         except AttributeError:
             tls = _callback_thread_local.tls = _ThreadLocal()
         return tls
-
-
-@contextlib.contextmanager
-def lock_directory(cache_dir, timeout=10, poll_interval=0.05):
-    """A simple lock to the callback cache directory.
-
-    This context manager prevents concurrent writing to the cache directory
-    by locking a file ".callback_lock" under the directory. Since the cuFFT
-    callbacks are only available in Linux, we use fcntl to achieve this.
-
-    This is essentially a stripped-down version of py-filelock.
-
-    Args:
-        cache_dir (str): path to the directory to be locked
-        timeout (float): maximal time in seconds to try to acquire the lock
-            before giving up (and raising a RuntimeError).
-        poll_interval (float): the time interval in seconds between two
-            consecutive attempts to acquire the lock.
-
-    """
-
-    lock_file = os.path.join(cache_dir, '.callback_lock')
-
-    def acquire(lock_file):
-        try:
-            fd = os.open(lock_file, os.O_RDWR | os.O_CREAT | os.O_TRUNC)
-            fcntl.flock(fd, fcntl.LOCK_EX | fcntl.LOCK_NB)
-        except (IOError, OSError):
-            os.close(fd)
-            fd = None
-        return fd
-
-    def release(fd):
-        fcntl.flock(fd, fcntl.LOCK_UN)
-        os.close(fd)
-
-    t_start = time.time()
-    fd = None
-    try:
-        while True:
-            fd = acquire(lock_file)
-            if fd is not None:
-                yield
-                break
-            elif time.time() - t_start > timeout:
-                raise RuntimeError('timeout is reached but a file lock still '
-                                   'cannot be acquired')
-            else:
-                time.sleep(poll_interval)
-    finally:  # exceptions can be raised from yield
-        if fd is not None:
-            release(fd)
 
 
 cdef inline void _set_vars() except*:
@@ -249,19 +194,23 @@ cdef inline void _mod_compile(str tempdir, str mod_name, str obj_host) except*:
     p.check_returncode()
 
 
-cdef inline str _prune(str cache_dir, str _cufft_ver, str arch):
-    cdef str cufft_lib_full, cufft_lib_pruned, cufft_lib_cached
+cdef inline str _prune(str temp_dir, str cache_dir, str _cufft_ver, str arch):
+    cdef str cufft_lib_full, cufft_lib_pruned, cufft_lib_temp, cufft_lib_cached
 
     if _nvprune:
         cufft_lib_full = os.path.join(_cuda_path, 'lib64/libcufft_static.a')
         cufft_lib_pruned = 'cufft_static_' + _cufft_ver + '_sm' + arch
+        cufft_lib_temp = os.path.join(temp_dir,
+                                      'lib' + cufft_lib_pruned + '.a')
         cufft_lib_cached = os.path.join(cache_dir,
                                         'lib' + cufft_lib_pruned + '.a')
         if not os.path.isfile(cufft_lib_cached):
-            with lock_directory(cache_dir):
-                p = subprocess.run([_nvprune, '-arch=sm_' + arch,
-                                    cufft_lib_full, '-o', cufft_lib_cached])
+            p = subprocess.run([_nvprune, '-arch=sm_' + arch,
+                                cufft_lib_full, '-o', cufft_lib_temp])
             p.check_returncode()
+            # atomic move with the destination guaranteed to be overwritten;
+            # using os.replace() is also ok here
+            os.rename(cufft_lib_temp, cufft_lib_cached)
     else:
         # nvprune is not found, just link against the full static lib
         cufft_lib_pruned = None
@@ -313,21 +262,26 @@ cdef inline void _nvcc_compile(
 
 
 cdef inline void _nvcc_link(
-        str tempdir, str obj_host, str obj_dev, str arch, str path,
+        str tempdir, str obj_host, str obj_dev, str arch, str mod_filename,
         str cache_dir, str cufft_lib_pruned) except*:
     # WARNING: CANNOT use host compiler to link!
     cdef list cmd
+    cdef str mod_temp, mod_cached
 
+    mod_temp = os.path.join(tempdir, mod_filename)
+    mod_cached = os.path.join(cache_dir, mod_filename)
     cmd = _nvcc + ['-shared', '-arch=sm_'+arch, obj_dev, obj_host]
     if cufft_lib_pruned:
         cmd += ['-L'+cache_dir, '-l'+cufft_lib_pruned]
     else:
         cmd.append('-lcufft_static')
-    cmd += ['-lculibos', '-lpthread', '-o', path]
+    cmd += ['-lculibos', '-lpthread', '-o', mod_temp]
 
-    with lock_directory(cache_dir):
-        p = subprocess.run(cmd, env=os.environ)
+    p = subprocess.run(cmd, env=os.environ)
     p.check_returncode()
+    # atomic move with the destination guaranteed to be overwritten;
+    # using os.replace() is also ok here
+    os.rename(mod_temp, mod_cached)
 
 
 cpdef get_current_callback_manager():
@@ -391,8 +345,9 @@ cdef class _CallbackManager:
             os.makedirs(cache_dir, exist_ok=True)
         path = os.path.join(cache_dir, mod_filename)
         if not os.path.isfile(path):
-            # Set up temp directory
-            tempdir_obj = tempfile.TemporaryDirectory()
+            # Set up temp directory; it must be under the cache directory so
+            # that atomic moves within the same filesystem can be guaranteed
+            tempdir_obj = tempfile.TemporaryDirectory(dir=cache_dir)
             tempdir = tempdir_obj.name
 
             # Cythonize the Cython code to produce a c++ source file
@@ -403,7 +358,8 @@ cdef class _CallbackManager:
             _mod_compile(tempdir, mod_name, obj_host)
 
             # Prune libcufft_static.a for the target arch and cache it
-            cufft_lib_pruned = _prune(cache_dir, str(_cufft_ver), arch)
+            cufft_lib_pruned = _prune(
+                tempdir, cache_dir, str(_cufft_ver), arch)
 
             # Dump and compile device code using nvcc
             obj_dev = os.path.join(tempdir, mod_name + '_dev.o')
@@ -411,7 +367,7 @@ cdef class _CallbackManager:
 
             # Use nvcc to link and generate a shared library, and place it in
             # the disk cache
-            _nvcc_link(tempdir, obj_host, obj_dev, arch, path,
+            _nvcc_link(tempdir, obj_host, obj_dev, arch, mod_filename,
                        cache_dir, cufft_lib_pruned)
 
             # Clean up build directory

--- a/cupy/fft/_callback.pyx
+++ b/cupy/fft/_callback.pyx
@@ -74,7 +74,7 @@ cdef class _ThreadLocal:
 
 
 @contextlib.contextmanager
-def lock_directory(timeout=10, poll_interval=0.05):
+def lock_directory(cache_dir, timeout=10, poll_interval=0.05):
     """A simple lock to the callback cache directory.
 
     This context manager prevents concurrent writing to the cache directory
@@ -84,6 +84,7 @@ def lock_directory(timeout=10, poll_interval=0.05):
     This is essentially a stripped-down version of py-filelock.
 
     Args:
+        cache_dir (str): path to the directory to be locked
         timeout (float): maximal time in seconds to try to acquire the lock
             before giving up (and raising a RuntimeError).
         poll_interval (float): the time interval in seconds between two
@@ -91,7 +92,7 @@ def lock_directory(timeout=10, poll_interval=0.05):
 
     """
 
-    lock_file = os.path.join(_callback_cache_dir, '.callback_lock')
+    lock_file = os.path.join(cache_dir, '.callback_lock')
 
     def acquire(lock_file):
         try:
@@ -253,7 +254,7 @@ cdef inline str _prune(str cache_dir, str _cufft_ver, str arch):
         cufft_lib_cached = os.path.join(cache_dir,
                                         'lib' + cufft_lib_pruned + '.a')
         if not os.path.isfile(cufft_lib_cached):
-            with lock_directory():
+            with lock_directory(cache_dir):
                 p = subprocess.run([_nvprune, '-arch=sm_' + arch,
                                     cufft_lib_full, '-o', cufft_lib_cached])
             p.check_returncode()
@@ -320,7 +321,7 @@ cdef inline void _nvcc_link(
         cmd.append('-lcufft_static')
     cmd += ['-lculibos', '-lpthread', '-o', path]
 
-    with lock_directory():
+    with lock_directory(cache_dir):
         p = subprocess.run(cmd, env=os.environ)
     p.check_returncode()
 

--- a/cupy/fft/_callback.pyx
+++ b/cupy/fft/_callback.pyx
@@ -6,6 +6,7 @@ from cupy.core.core cimport ndarray
 from cupy.cuda.device cimport get_compute_capability
 
 import contextlib
+import fcntl
 import hashlib
 import importlib
 import os
@@ -16,6 +17,7 @@ import sys
 import sysconfig
 import tempfile
 import threading
+import time
 import warnings
 
 from cupy import __version__ as _cupy_ver
@@ -69,6 +71,53 @@ cdef class _ThreadLocal:
         except AttributeError:
             tls = _callback_thread_local.tls = _ThreadLocal()
         return tls
+
+
+@contextlib.contextmanager
+def lock_directory(timeout=10, poll_interval=0.05):
+    """A simple lock to the callback cache directory.
+
+    This context manager prevents concurrent writing to the cache directory
+    by locking a file ".callback_lock" under the directory. Since the cuFFT
+    callbacks are only available in Linux, we use fcntl to achieve this.
+
+    This is essentially a stripped-down version of py-filelock.
+
+    Args:
+        timeout (float): maximal time in seconds to try to acquire the lock
+            before giving up (and raising a RuntimeError).
+        poll_interval (float): the time interval in seconds between two
+            consecutive attempts to acquire the lock.
+
+    """
+
+    lock_file = os.path.join(_callback_cache_dir, '.callback_lock')
+
+    def acquire(lock_file):
+        try:
+            fd = os.open(lock_file, os.O_RDWR | os.O_CREAT | os.O_TRUNC)
+            fcntl.flock(fd, fcntl.LOCK_EX | fcntl.LOCK_NB)
+        except (IOError, OSError):
+            os.close(fd)
+            fd = None
+        return fd
+
+    def release(fd):
+        fcntl.flock(fd, fcntl.LOCK_UN)
+        os.close(fd)
+
+    t_start = time.time()
+    while True:
+        fd = acquire(lock_file)
+        if fd is not None:
+            yield
+            break
+        elif time.time() - t_start > timeout:
+            raise RuntimeError('timeout is reached but a file lock still '
+                               'cannot be acquired')
+        else:
+            time.sleep(poll_interval)
+    release(fd)
 
 
 cdef inline void _set_vars() except*:
@@ -204,8 +253,9 @@ cdef inline str _prune(str cache_dir, str _cufft_ver, str arch):
         cufft_lib_cached = os.path.join(cache_dir,
                                         'lib' + cufft_lib_pruned + '.a')
         if not os.path.isfile(cufft_lib_cached):
-            p = subprocess.run([_nvprune, '-arch=sm_' + arch,
-                                cufft_lib_full, '-o', cufft_lib_cached])
+            with lock_directory():
+                p = subprocess.run([_nvprune, '-arch=sm_' + arch,
+                                    cufft_lib_full, '-o', cufft_lib_cached])
             p.check_returncode()
     else:
         # nvprune is not found, just link against the full static lib
@@ -270,7 +320,8 @@ cdef inline void _nvcc_link(
         cmd.append('-lcufft_static')
     cmd += ['-lculibos', '-lpthread', '-o', path]
 
-    p = subprocess.run(cmd, env=os.environ)
+    with lock_directory():
+        p = subprocess.run(cmd, env=os.environ)
     p.check_returncode()
 
 
@@ -280,8 +331,6 @@ cpdef get_current_callback_manager():
     return mgr
 
 
-# TODO(leofang): find a way to implement a lock-free method for
-# cached shared libraries like what's done in cupy/cuda/compiler.py?
 cdef class _CallbackManager:
     cdef:
         readonly str cb_load
@@ -511,12 +560,6 @@ cdef class set_cufft_callbacks:
         load/store callbacks). Due to static linking, however, the file sizes
         can be excessive! The cache position can be changed via setting
         ``CUPY_CACHE_DIR``.
-
-    .. warning::
-        This feature may not be thread- or process-safe. In a concurrent
-        environment, it is advised to first let the main thread/process do the
-        warm-up (compiling), and then unblock all threads/processes to load the
-        compiled module from cache.
 
     .. seealso:: `cuFFT Callback Routines`_
 

